File: src/lib/server/logging.ts
import type { RequestEvent } from '@sveltejs/kit'

export function logAccess(input: { event: RequestEvent; action: string; details?: Record<string, unknown> }): void {
  const { event, action, details = {} } = input
  const ip = getRequestIp(event)
  logEvent({ ip, action, details })
}

export function logEvent(input: { ip: string; action: string; details?: Record<string, unknown> }): void {
  const { ip, action, details = {} } = input
  const timestamp = new Date().toISOString()
  const defaultLevel = action.endsWith('_error') ? 'ERROR' : 'INFO'
  const { level = defaultLevel, ...rest } = details
  const serializedDetails = Object.entries(rest)
    .map(([key, value]) => `${key}=${JSON.stringify(value)}`)
    .join(' ')
  console.log(
    `${timestamp} ${level} ip=${ip || 'unknown'} action=${action}${serializedDetails ? ` ${serializedDetails}` : ''}`,
  )
}

function getRequestIp(event: { request: Request; getClientAddress: () => string }): string {
  const { request, getClientAddress } = event
  try {
    return getClientAddress()
  } catch {
    const forwardedFor = request.headers.get('x-forwarded-for')
    if (forwardedFor) {
      return forwardedFor.split(',')[0].trim()
    }
    return 'unknown'
  }
}

File: backend/app/logging.py
from __future__ import annotations

import json
from datetime import datetime, timezone

LEVEL_ACTIONS_ERROR = "_error"


def log_event(*, ip: str, action: str, details: dict | None = None) -> None:
    """Log an event in the same line format as the front-end app.

    Format: `<timestamp> <level> ip=<ip|unknown> action=<action> key="value" ...`
    `level` defaults to ERROR for `*_error` actions, else INFO; an explicit
    `level` key in `details` overrides it.
    """
    details = dict(details or {})
    timestamp = (
        datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
    )
    default_level = "ERROR" if action.endswith(LEVEL_ACTIONS_ERROR) else "INFO"
    level = details.pop("level", default_level)
    serialized = " ".join(f"{key}={json.dumps(value)}" for key, value in details.items())
    line = f"{timestamp} {level} ip={ip or 'unknown'} action={action}"
    if serialized:
        line += f" {serialized}"
    print(line)


def get_request_ip(request) -> str:
    forwarded_for = request.headers.get("x-forwarded-for")
    if forwarded_for:
        return forwarded_for.split(",")[0].strip()
    if request.client is not None:
        return request.client.host
    return "unknown"